src/bin/pg_autoctl/keeper.c

```diff
@@ -947,6 +946,32 @@ keeper_ensure_configuration(Keeper *keeper, bool postgresNotRunningIsOk)
 			}
 		}
 	}
+	else if (state->current_role == SINGLE_STATE ||
+			 state->current_role == WAIT_PRIMARY_STATE ||
+			 state->current_role == JOIN_PRIMARY_STATE ||
+			 state->current_role == PRIMARY_STATE ||
+			 state->current_role == APPLY_SETTINGS_STATE)
+	{
+		/*
+		 * On the primary, the keeper's replication password may have been set
+		 * in the keeper config while the server was stopped. So, set the
+		 * password in the server catalog.
+		 */
```

Comment:
It would be better to only call `ALTER ROLE ... PASSWORD` in the three following situations:

  - when we start Postgres for the first time in `pg_autoctl` (because it might have been changed at rest),
  - or the first time after an interactive password change,
  - or any time we check that Postgres is running if we failed to update the password on the previous occasion.

To enable that,  we could have a cache of the password in-memory. The cache should probably live in the `LocalPostgresServer` data structure, invalidated at reload time (first "reload" to set the value, when the cache is unset, and we know that thanks to IS_EMPTY_STRING_BUFFER, then only when different from the existing value).

It looks like the `pgsql_set_password` call could then happen in `ensure_postgres_service_is_running` in the case when the function had to actually restart Postgres, and that an update is expected.  Because `ensure_postgres_service_is_running` only has access to the `LocalPostgresServer` rather than the whole `Keeper`, we might need to also maintain a boolean in `LocalPostgresServer ` to drive the decision to call to `pgsql_set_password`. The boolean would be maintained in the `keeper_ensure_configuration` when we also update our password cache.

Note that we would still keep retrying to set the password an infinite amount of time with that scheme, though it's now easy to solve that problem, adding a counter to the password change tracking boolean. We could decide that after N password change attempts we bail out, but I'm not sure what we should do in that case. The user certainly needs to find another password that Postgres will like better. At least the `pgsql_set_password` retry mechanism would be separate from the reload signal handling mechanism and would not make it so that we keep reloading the whole setup again and again. When someone is in the middle of editing their pg_auto_failover setup, it's best to not have rewrites of the configuration file happening in the background.

Reply:
In that instance I specifically meant `ensure_postgres_service_is_running` as written later.